File: CSharp/SpaceVIL/SpaceVIL/Items/ToolTipItem.cs
using System;
using System.Drawing;
using System.Timers;
using SpaceVIL.Core;
using SpaceVIL.Common;
using SpaceVIL.Decorations;
using System.Collections.Generic;

namespace SpaceVIL
{
    public sealed class ToolTipItem : Prototype, IFloating
    {
        private Label _textObject;
        internal Label GetTextLabel()
        {
            return _textObject;
        }
        internal Timer _stop;
        private int _timeout = 500;
        internal void SetTimeOut(int milliseconds)
        {
            _timeout = milliseconds;
        }
        internal int GetTimeOut()
        {
            return _timeout;
        }

        // private static ToolTip _instance = null;

        internal ToolTipItem()
        {
            SetVisible(false);
            _textObject = new Label();
            SetItemName("ToolTip");
            IsFocusable = false;
            _queue = new Queue<IBaseItem>();
            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.ToolTipItem)));
        }

        internal bool IsInit = false;

        public override void InitElements()
        {
            ItemsLayoutBox.AddItem(GetHandler(), this, LayoutType.Floating);
            AddItem(_textObject);
            if (_queue.Count != 0)
            {
                foreach (IBaseItem item in _queue)
                {
                    base.AddItem(item);
                }
            }
            IsInit = true;
        }

        private Queue<IBaseItem> _queue = null;

        public override void AddItems(params IBaseItem[] items)
        {
            if (IsInit)
                return;
            foreach (IBaseItem item in items)
            {
                _queue.Enqueue(item);
            }
        }

        internal void InitTimer(bool value)
        {
            if (value)
            {
                if (_stop != null)
                    return;

                if (_timeout == 0)
                    SetVisible(true);
                else
                {

                    _stop = new System.Timers.Timer(_timeout);
                    _stop.Elapsed += (sender, e) => VisibleSelf();
                    _stop.Start();
                }
            }
            else
            {
                SetVisible(false);

                if (_stop == null)
                    return;

                _stop.Stop();
                _stop.Dispose();
                _stop = null;
            }
        }

        private void VisibleSelf()
        {
            SetVisible(true);

            _stop.Stop();
            _stop.Dispose();
            _stop = null;
        }

        internal void SetTextAlignment(ItemAlignment alignment)
        {
            _textObject.SetTextAlignment(alignment);
        }
        internal void SetTextAlignment(params ItemAlignment[] alignment)
        {
            _textObject.SetTextAlignment(alignment);
        }
        internal void SetFont(Font font)
        {
            _textObject.SetFont(font);
        }
        internal void SetFontSize(int size)
        {
            _textObject.SetFontSize(size);
        }
        internal void SetFontStyle(FontStyle style)
        {
            _textObject.SetFontStyle(style);
        }
        internal void SetFontFamily(FontFamily font_family)
        {
            _textObject.SetFontFamily(font_family);
        }
        internal Font GetFont()
        {
            return _textObject.GetFont();
        }
        internal void SetText(String text)
        {
            _textObject.SetText(text);
        }
        internal String GetText()
        {
            return _textObject.GetText();
        }
        internal int GetTextWidth()
        {
            return _textObject.GetTextWidth();
        }
        internal int GetTextHeight()
        {
            return _textObject.GetTextHeight();
        }
        internal void SetForeground(Color color)
        {
            _textObject.SetForeground(color);
        }
        internal Color GetForeground()
        {
            return _textObject.GetForeground();
        }

        //style
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;
            base.SetStyle(style);

            Style innerStyle = style.GetInnerStyle("text");
            if (innerStyle != null)
                _textObject.SetStyle(innerStyle);

            SetForeground(style.Foreground);
            SetFont(style.Font);
            SetTextAlignment(style.TextAlignment);
        }

        public void Show()
        {
        }

        public void Show(IItem sender, MouseArgs args)
        {
        }

        public void Hide()
        {
        }

        public void Hide(MouseArgs args)
        {
        }

        public bool IsOutsideClickClosable()
        {
            return false;
        }

        public void SetOutsideClickClosable(bool value)
        {
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/ToolTipItem.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Core.InterfaceBaseItem;
import com.spvessel.spacevil.Core.InterfaceFloating;
import com.spvessel.spacevil.Core.InterfaceItem;
import com.spvessel.spacevil.Core.MouseArgs;
import com.spvessel.spacevil.Decorations.Indents;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.ItemAlignment;
import com.spvessel.spacevil.Flags.LayoutType;

import java.awt.Font;
import java.awt.Color;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.ArrayDeque;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import javax.swing.Timer;

public final class ToolTipItem extends Prototype implements InterfaceFloating {

    private Label _textObject;

    Label getTextLabel() {
        return _textObject;
    }

    private Timer _stop = null;
    private int _timeout = 500;

    void setTimeOut(int milliseconds) {
        _timeout = milliseconds;
    }

    int getTimeOut() {
        return _timeout;
    }

    // private static ToolTip _instance = null;

    protected ToolTipItem() {
        setVisible(false);
        _textObject = new Label();
        setItemName("ToolTipItem");
        _textObject.setItemName("ToolTipLabel");
        isFocusable = false;
        _queue = new ArrayDeque<>();
        setStyle(DefaultsService.getDefaultStyle(ToolTipItem.class));
    }

    boolean isInit = false;

    @Override
    public void initElements() {
        ItemsLayoutBox.addItem(getHandler(), this, LayoutType.FLOATING);
        addItem(_textObject);
        if (!_queue.isEmpty()) {
            for (InterfaceBaseItem item : _queue) {
                super.addItem(item);
            }
        }
        isInit = true;
    }

    private Queue<InterfaceBaseItem> _queue = null;

    @Override
    public void addItems(InterfaceBaseItem... items) {
        if (isInit)
            return;
        for (InterfaceBaseItem item : items) {
            _queue.add(item);
        }
    }

    private Lock locker = new ReentrantLock();

    void initTimer(boolean value) {
        locker.lock();
        try {
            if (value) {
                if (_stop != null)
                    return;
                ActionListener taskPerformer = new ActionListener() {
                    public void actionPerformed(ActionEvent evt) {
                        visibleSelf();
                    }
                };
                _stop = new Timer(_timeout, taskPerformer);
                _stop.start();
            } else {
                setVisible(value);
                if (_stop == null)
                    return;
                _stop.stop();
                _stop = null;
            }
        } finally {
            locker.unlock();
        }
    }

    private void visibleSelf() {
        locker.lock();
        try {
            if (_stop == null)
                return;
            setVisible(true);
            _stop.stop();
            _stop = null;
        } finally {
            locker.unlock();
        }
    }

    int getTextWidth() {
        return _textObject.getTextWidth();
    }

    int getTextHeight() {
        return _textObject.getTextHeight();
    }

    // text init
    void setTextAlignment(List<ItemAlignment> alignment) {
        _textObject.setTextAlignment(alignment);
    }

    void setTextMargin(Indents margin) {
        _textObject.setMargin(margin);
    }

    void setFont(Font font) {
        _textObject.setFont(font);
    }

    void setFontSize(int size) {
        _textObject.setFontSize(size);
    }

    void setFontStyle(int style) {
        _textObject.setFontStyle(style);
    }

    void setFontFamily(String font_family) {
        _textObject.setFontFamily(font_family);
    }

    Font getFont() {
        return _textObject.getFont();
    }

    void setText(String text) {
        _textObject.setText(text);
    }

    String getText() {
        return _textObject.getText();
    }

    void setForeground(Color color) {
        _textObject.setForeground(color);
    }

    void setForeground(int r, int g, int b) {
        _textObject.setForeground(r, g, b);
    }

    void setForeground(int r, int g, int b, int a) {
        _textObject.setForeground(r, g, b, a);
    }

    void setForeground(float r, float g, float b) {
        _textObject.setForeground(r, g, b);
    }

    void setForeground(float r, float g, float b, float a) {
        _textObject.setForeground(r, g, b, a);
    }

    Color getForeground() {
        return _textObject.getForeground();
    }

    @Override
    public void setStyle(Style style) {
        if (style == null)
            return;
        super.setStyle(style);

        Style innerStyle = style.getInnerStyle("text");
        if (innerStyle != null)
            _textObject.setStyle(innerStyle);

        setForeground(style.foreground);
        setFont(style.font);
        setTextAlignment(style.textAlignment);
    }

    @Override
    public void show() {
    }

    @Override
    public void show(InterfaceItem sender, MouseArgs args) {
    }

    @Override
    public void hide() {
    }

    @Override
    public void hide(MouseArgs args) {
    }

    @Override
    public boolean isOutsideClickClosable() {
        return false;
    }

    @Override
    public void setOutsideClickClosable(boolean value) {
    }
}
